internal/oauth2/lazy_token.go

```diff
@@ -0,0 +1,105 @@
+package oauth2
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Tokener is an interface that defines the AccessToken method.
+type Tokener interface {
+	NextToken(context.Context) (string, error)
+	ForceRefresh(context.Context) (string, error)
+}
+
+type LazyToken struct {
+	// Url represents the URL used for acquiring the token.
+	Url string
+	// ClientId is the client ID used for authentication.
+	ClientId string
+	// ClientSecret is the client secret used for authentication.
+	ClientSecret string
+	// 	AccessToken string holds the currently cached token.
+	AccessToken string
+	// 	RefreshToken string token after refresh.
+	RefreshToken string
+	// Expiration stores the expiration time of the current token.
+	Expiration time.Time
+	// mutex ensures safe concurrent access to the token and expiration fields.
+	mutex sync.Mutex
+}
+
+type tokenResponse struct {
+	AccessToken string    `json:"access_token"`
+	Expiration  time.Time `json:"expiration"`
+}
+
+// NextToken returns a valid token, acquiring a new one if necessary.
+func (lt *LazyToken) NextToken(ctx context.Context) (string, error) {
+	lt.mutex.Lock()
+	defer lt.mutex.Unlock()
+
+	if lt.AccessToken == "" || time.Now().After(lt.Expiration.Add(-time.Minute)) {
+		// Acquire a new token and set the expiration time
+		tokenRes, err := lt.requestToken()
+		if err != nil {
+			return "", err
+		}
+
+		lt.AccessToken = tokenRes.AccessToken
+		lt.Expiration = tokenRes.Expiration
+		logrus.WithContext(ctx).Infof("Acquired new token %s with expiration at: %s", lt.AccessToken, lt.Expiration)
+	} else {
+		logrus.WithContext(ctx).Infof("AccessToken requested: %s (expires at %s)", lt.AccessToken, lt.Expiration)
+	}
+
+	return lt.AccessToken, nil
+}
+
+func (lt *LazyToken) ForceRefresh(ctx context.Context) (string, error) {
+	lt.AccessToken = ""
```

Comment:
This is a change which is not protected by the `mutex`, generally this is a bad practice but in this case if I am not mistaken it is not an issue since blank access token means it is gonna be refreshed on next update.

Reply:
yes, it will refresh on next update, I think it is what we want here

Reply:
I think this is actually racy. Imagine `ForceRefresh` being called before this return statement: https://github.com/mgold1234/image-builder/blob/20d8326200dd50a030037bd660bb21c7589aa9d2/internal/oauth2/lazy_token.go#L65

So I guess we have to extract

```
if lt.AccessToken == "" || time.Now().Add(time.Minute).After(lt.Expiration) {
		// Acquire a new token and set the expiration time
		tokenRes, err := lt.requestToken()
		if err != nil {
			return "", err
		}

		lt.AccessToken = tokenRes.AccessToken
		lt.Expiration = tokenRes.Expiration
		lt.RefreshToken = tokenRes.RefreshToken
		logrus.WithContext(ctx).Infof("Acquired new token %s with expiration at: %s", lt.AccessToken, lt.Expiration)
	} else {
		logrus.WithContext(ctx).Infof("AccessToken reused: %s (expires at %s)", lt.AccessToken, lt.Expiration)
	}

	return lt.AccessToken, nil
```

to a separate private function, and protect both `RefreshToken` and `ForceRefresh` with the same mutex, and let them call the private function.

Reply:
Ah right my guts were right, I missed the window between setting the token and returning the statement. Yeah, good point, this needs to be extracted.